File: 026_Merge_Sorted_Arrays_Next_Permutation/next_permutation.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    void nextPermutation(vector<int>& A) {
        int n = A.size();
        int pivot = -1;

        // Step 1: find the pivot
        for (int i = n - 2; i >= 0; i--) { // n-2 because we check A[i] < A[i+1]
            if (A[i] < A[i + 1]) {
                pivot = i;
                break;
            }
        }

        // If no pivot found, reverse the array
        if (pivot == -1) {
            reverse(A.begin(), A.end());
            return;
        }

        // Step 2: find the next larger element to swap with pivot
        for (int i = n - 1; i > pivot; i--) {
            if (A[i] > A[pivot]) {
                swap(A[i], A[pivot]);
                break;
            }
        }

        // Step 3: reverse the elements after pivot
        int i = pivot + 1, j = n - 1;
        while (i < j) {
            swap(A[i], A[j]);
            i++;
            j--;
        }
    }
};

int main() {
    Solution sol;
    vector<int> nums = {1, 2, 3};
    
    sol.nextPermutation(nums);

    cout << "Next permutation: ";
    for (int num : nums) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: 026_Merge_Sorted_Arrays_Next_Permutation/next_permutation.py
def nextPermutation(nums):
    n = len(nums)
    pivot = -1

    # Step 1: find the pivot (first number from right that is smaller than its next)
    for i in range(n - 2, -1, -1):
        if nums[i] < nums[i + 1]:
            pivot = i
            break

    # If no pivot found, the array is in descending order, so reverse it
    if pivot == -1:
        nums.reverse()
        return

    # Step 2: find the next larger element to swap with pivot
    for i in range(n - 1, pivot, -1):
        if nums[i] > nums[pivot]:
            nums[i], nums[pivot] = nums[pivot], nums[i]
            break

    # Step 3: reverse the part after pivot
    nums[pivot + 1:] = reversed(nums[pivot + 1:])

# Example usage
nums = [1, 2, 3]
nextPermutation(nums)
print("Next permutation:", nums)